### PR TITLE
Implemented Metis mesh partitioner

### PR DESCRIPTION
**Pull request by biancagi**
Implemented mesh partitioning based on `metis`. All `mito` functionalities can now run with `MPI` multiprocessing on partitioned meshes.

All added features are tested in the test suite.

### CODE REVIEW

File: lib/mito/mesh/metis/Partitioner.icc [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+template <class meshT>
+auto
+mito::mesh::metis::Partitioner<meshT>::_populate_element_connectivity(
+    const auto & cells, const auto & vertex_to_id,
+    const mito::geometry::geometry_t<mesh_t::dim> & geometry) -> auto
+{
+    // get the total number of cells in the mesh
+    int n_elements = cells.size();
```
Comment (akoshakji) [inline lib/mito/mesh/metis/Partitioner.icc:75]: please use `std::size`

File: lib/mito/mesh/metis/Partitioner.icc [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    auto cells = _get_ordered_mesh_cells(mesh);
+
+    // populate elements connectivity
+    auto connectivity = _populate_element_connectivity(cells, vertex_to_id, mesh.geometry());
+
+    // get the total number of cells in the mesh
+    int n_elements = cells.size();
+
+    // get the total number of vertices in the mesh
+    int n_vertices = vertex_to_id.size();
```
Comment (akoshakji) [inline lib/mito/mesh/metis/Partitioner.icc:182]: please use std::size

File: lib/mito/topology/OrientedSimplex.h [hunk 3]
```diff
@@ -74,14 +77,47 @@ namespace mito::topology {
...
+            vertex_set_t vertices_collection;
+
+            // fetch the vertices of this simplex
+            _footprint->vertices(vertices_collection);
+
+            // an array to store the composition of a D-simplex in terms of vertices
+            vertex_simplex_composition_t<D> vertices;
+
+            // assert that you found the correct number of vertices
+            assert(vertices_collection.size() == vertices.size());
```
Comment (akoshakji) [inline lib/mito/topology/OrientedSimplex.h:101]: please use std::size

File: lib/mito/topology/OrientedSimplex.h [hunk 4]
```diff
@@ -74,14 +77,47 @@ namespace mito::topology {
...
+
+            // populate the array of vertices
+            int i = 0;
+            for (auto & vertex : vertices_collection) {
```
Comment (akoshakji) [inline lib/mito/topology/OrientedSimplex.h:105]: `vertex` could be `const`

File: tests/mito.lib/mesh/metis_partitioner.cc [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    // rank of the mesh to return
+    int n_rank = 0;
+
+    // partition the mesh
+    auto mesh_partition = mito::mesh::metis::partition(mesh, n_partitions, n_rank);
+
+    // expect that the mesh was partitioned equally
+    // (this check assumes that the number of cells of the original mesh is divisible by the number
+    //  of partitions requested)
+    EXPECT_EQ(mesh_partition.cells().size(), mesh.cells().size() / n_partitions);
```
Comment (akoshakji) [inline tests/mito.lib/mesh/metis_partitioner.cc:115]: please use std::size

**Review by akoshakji — COMMENTED** [review]
minor comments, ready for merge after those are addressed

File: lib/mito/geometry/utilities.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// code guard
+#if !defined(mito_geometry_utilities_h)
+#define mito_geometry_utilities_h
+
+
+namespace mito::geometry {
+
+    // compute the barycenter of a cell, given a geometry
+    template <class cellT, int D>
+    auto barycenter(const cellT & cell, const geometry_t<D> & geometry) -> vector_t<D>
```
Comment (biancagi) [inline lib/mito/geometry/utilities.h:10]: This function needs to be tested in a unit test.

File: lib/mito/mesh/metis/Partitioner.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+      private:
+        // populate a map between vertices and a continuous integer id = 0, ..., n_vertices - 1
+        static auto _populate_vertices_map(const mesh_t & mesh) -> auto;
+        // get the mesh cells in with a lexicographic order on the barycenter
+        static auto _get_ordered_mesh_cells(const mesh_t & mesh) -> auto;
+        // populate elements connectivity
+        static auto _populate_element_connectivity(
+            const auto & cells, const auto & vertex_to_id,
+            const mito::geometry::geometry_t<mesh_t::dim> & geometry) -> auto;
```
[the remark this answers is not in the document]
Comment (biancagi) [reply lib/mito/mesh/metis/Partitioner.h:27]: I agree. That will require some class level `typedef`s to make things more readable.

File: lib/mito/topology/OrientedSimplex.h [hunk 8]
```diff
@@ -74,14 +77,47 @@ namespace mito::topology {
...
+                ++i;
+            }
+
+            // return the array of vertices
+            return vertices;
+        }
+
+        // append the edges of this simplex to a collection of edges
+        template <class EDGES_COLLECTION_T>
+        inline auto edges(EDGES_COLLECTION_T & edges) const -> void
```
Comment (biancagi) [inline lib/mito/topology/OrientedSimplex.h:116]: This method requires a unit test.

**Review by akoshakji — APPROVED** [review]

**Pull request merged**